### PR TITLE
Improved WithCargo performance

### PR DESCRIPTION
**Pull request by Mailaender**
Closes https://github.com/OpenRA/OpenRA/issues/9183.

Also moved it to Mods.Common as this can be used everywhere.

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/Render/WithCargo.cs [hunk 1]
```diff
@@ -29,13 +28,16 @@ public class WithCargoInfo : ITraitInfo, Requires<CargoInfo>, Requires<IBodyOrie
...
-	public class WithCargo : IRenderModifier, ITick
+	public class WithCargo : IRenderModifier, ITick, INotifyPassengerEntered, INotifyPassengerExited
 	{
 		readonly Cargo cargo;
 		readonly IFacing facing;
 		readonly WithCargoInfo cargoInfo;
 		readonly IBodyOrientation body;
 
+		List<IFacing> passengerFacings;
+		Dictionary<Actor, PassengerInfo> passengerInfos;
```
Comment (RoosterDragon) [inline OpenRA.Mods.Common/Traits/Render/WithCargo.cs]: ```readonly``` please.

**Comment by reaperrr** [comment]
Needs rebase.

**Comment by Mailaender** [comment]
Rebased and updated.

**Comment by RoosterDragon** [comment]
:+1: Looks good to me.

**Comment by reaperrr** [comment]
Needs rebase.

**Comment by Mailaender** [comment]
Rebased.

**Comment by penev92** [comment]
This causes some very strange passenger behavior when unloading, which is quite visible on the first GDI mission.

**Comment by reaperrr** [comment]
Also needs rebase now.

**Comment by Mailaender** [comment]
Sadly this regresses hilariously and I am not sure why it happens.

**Comment by obrakmann** [comment]
The reason for the regression is that the cargo is added twice to the `passengerFacings` (and `passengerInfos`, probably) list. If you create a transport actor and then immediately load cargo into it, `PassengerEntered` is called twice, once from `Cargo.Tick` (because `initialized` is still false), and once from `Cargo.Load`. Setting `initialized` to true in `Cargo.Load` would be one way to fix that, another more correct way would be to use `Cargo.Load` in the constructor (otherwise potential upgrades provided by the cargo wouldn't be applied).

**Comment by pchote** [comment]
passengerFacings probably wants to be a set rather than a list, independent of the other fixes.

**Comment by Mailaender** [comment]
Doesn't really matter anyway as `WithCargo` isn't used extensively.

**Pull request closed without merging**